src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs

```diff
@@ -410,307 +409,143 @@ public void FormulaCantRedefine()
+            true,
+            null)]
+        [InlineData(
+            "foo(x:Number):Number = x + 1;",
+            "foo(Float(1))",
+            false,
+            2.0)]
+
+        public void UserDefiniedFunctionTest(string udfExpression, string expression, bool expectedError, object expected, bool isDecimal = false)
         {
             var config = new PowerFxConfig()
             {
                 MaxCallDepth = 100
             };
             var recalcEngine = new RecalcEngine(config);
-            var body = @"If(Not(x = 1), If(Mod(x, 2)=0, hailstone(x/2), hailstone(3*x+1)), x)";
-
-            Assert.False(recalcEngine.DefineFunctions($"hailstone(x:Number):Number = {body};").Errors.Any());
-            Assert.Equal(1.0, recalcEngine.Eval("hailstone(Float(192))").ToObject());
-        }
 
-        [Fact]
-        public void DefHailstoneSequenceDecimal()
-        {
-            var config = new PowerFxConfig()
+            try
             {
-                MaxCallDepth = 100
-            };
-            var recalcEngine = new RecalcEngine(config);
-            var body = @"If(Not(x = 1), If(Mod(x, 2)=0, hailstone(x/2), hailstone(3*x+1)), x)";
+                recalcEngine.AddUserDefinedFunction(udfExpression, CultureInfo.InvariantCulture);
 
-            Assert.False(recalcEngine.DefineFunctions($"hailstone(x:Decimal):Decimal = {body};").Errors.Any());
-            Assert.Equal(1m, recalcEngine.Eval("hailstone(192)").ToObject());
-        }
+                var check = recalcEngine.Check(expression);
 
-        [Fact]
-        public void DefMutualRecursionFunc()
-        {
-            var config = new PowerFxConfig()
-            {
-                MaxCallDepth = 100
-            };
-            var recalcEngine = new RecalcEngine(config);
-            var bodyEven = @"If(number = 0, true, odd(Abs(number)-1))";
-            var bodyOdd = @"If(number = 0, false, even(Abs(number)-1))";
+                Assert.Equal(check.IsSuccess, !expectedError);
 
-            var opts = new ParserOptions() { NumberIsFloat = true };
-            Assert.False(recalcEngine.DefineFunctions($"odd(number:Number):Boolean = {bodyOdd}; even(number:Number):Boolean = {bodyEven};").Errors.Any());
-            Assert.Equal(true, recalcEngine.Eval("odd(17)", options: opts).ToObject());
-            Assert.Equal(false, recalcEngine.Eval("even(17)", options: opts).ToObject());
-        }
+                var result = recalcEngine.Eval(expression);
 
-        [Fact]
-        public void DefMutualRecursionFuncDecimal()
-        {
-            var config = new PowerFxConfig()
+                if (isDecimal)
+                {
+                    var decimalValue = decimal.Parse(expected.ToString());
+                    Assert.Equal(decimalValue, result.ToObject());
+                }
+                else
+                {
+                    Assert.Equal(expected, result.ToObject());
+                }
+            }
+            catch (Exception ex)

```

Comment:
>catch (Exception ex) [](http://example.com/codeflow?start=12&length=20)

future PR: we should be able to detect user errors without having to catch an exception (like CheckResult.IsSuccess)